File: src/codegen/alembic/migration.ts
import type {
  DatabaseSchema,
  TableSpec,
  ColumnSpec,
} from "#convention/types.js";

export interface AlembicColumn {
  readonly name: string;
  readonly saType: string;
  readonly nullable: boolean;
  readonly primaryKey: boolean;
  readonly autoincrement: boolean;
  readonly serverDefault: string | null;
}

export interface AlembicForeignKey {
  readonly name: string;
  readonly column: string;
  readonly refTable: string;
  readonly refColumn: string;
  readonly onDelete: string;
}

export interface AlembicCheck {
  readonly name: string;
  readonly sql: string;
}

export interface AlembicIndex {
  readonly name: string;
  readonly table: string;
  readonly columns: readonly string[];
  readonly unique: boolean;
}

export interface AlembicTable {
  readonly name: string;
  readonly entityName: string;
  readonly columns: readonly AlembicColumn[];
  readonly foreignKeys: readonly AlembicForeignKey[];
  readonly checks: readonly AlembicCheck[];
  readonly indexes: readonly AlembicIndex[];
  readonly tableArgs: readonly string[];
}

export interface AlembicMigration {
  readonly revision: string;
  readonly createdDate: string;
  readonly tables: readonly AlembicTable[];
  readonly tablesReversed: readonly AlembicTable[];
  readonly needsPostgresDialect: boolean;
}

export interface BuildMigrationOptions {
  readonly revision?: string;
  readonly createdDate?: string;
}

export function buildAlembicMigration(
  schema: DatabaseSchema,
  opts: BuildMigrationOptions = {},
): AlembicMigration {
  const sorted = topoSortTables(schema.tables);
  const tables = sorted.map((t) => buildAlembicTable(t));
  const needsPostgresDialect = tables.some((t) =>
    t.columns.some((c) => c.saType.startsWith("postgresql.")),
  );
  return {
    revision: opts.revision ?? "001",
    createdDate: opts.createdDate ?? new Date().toISOString().slice(0, 10),
    tables,
    tablesReversed: [...tables].reverse(),
    needsPostgresDialect,
  };
}

function topoSortTables(tables: readonly TableSpec[]): TableSpec[] {
  const byName = new Map(tables.map((t) => [t.name, t]));
  const visited = new Set<string>();
  const result: TableSpec[] = [];

  function visit(t: TableSpec): void {
    if (visited.has(t.name)) return;
    visited.add(t.name);
    for (const fk of t.foreignKeys) {
      const target = byName.get(fk.refTable);
      if (target !== undefined && target !== t) visit(target);
    }
    result.push(t);
  }
  for (const t of tables) visit(t);
  return result;
}

function buildAlembicTable(t: TableSpec): AlembicTable {
  const columns = t.columns.map((c) => buildColumn(c, t));
  const foreignKeys = t.foreignKeys.map((fk) => ({
    name: `fk_${t.name}_${fk.column}`,
    column: fk.column,
    refTable: fk.refTable,
    refColumn: fk.refColumn,
    onDelete: fk.onDelete,
  }));
  const checks = t.checks.map((sql, i) => ({
    name: `ck_${t.name}_${i}`,
    sql,
  }));
  const indexes = t.indexes.map((ix) => ({
    name: ix.name,
    table: t.name,
    columns: ix.columns,
    unique: ix.unique,
  }));
  const tableArgs = [
    ...columns.map(renderColumnCall),
    ...foreignKeys.map(renderForeignKeyCall),
    ...checks.map(renderCheckCall),
  ];
  return {
    name: t.name,
    entityName: t.entityName,
    columns,
    foreignKeys,
    checks,
    indexes,
    tableArgs,
  };
}

function buildColumn(c: ColumnSpec, t: TableSpec): AlembicColumn {
  const isPk = c.name === t.primaryKey;
  const isSerial = c.sqlType === "BIGSERIAL" || c.sqlType === "SERIAL";
  return {
    name: c.name,
    saType: mapSqlTypeToSa(c.sqlType),
    nullable: c.nullable,
    primaryKey: isPk,
    autoincrement: isSerial,
    serverDefault: mapServerDefault(c.defaultValue),
  };
}

function mapSqlTypeToSa(sqlType: string): string {
  const direct = DIRECT_SA_TYPES.get(sqlType);
  if (direct !== undefined) return direct;
  const numericMatch = sqlType.match(/^NUMERIC\((\d+)\s*,\s*(\d+)\)$/);
  if (numericMatch) return `sa.Numeric(${numericMatch[1]}, ${numericMatch[2]})`;
  const varcharMatch = sqlType.match(/^VARCHAR\((\d+)\)$/);
  if (varcharMatch) return `sa.String(length=${varcharMatch[1]})`;
  return "sa.Text()";
}

const DIRECT_SA_TYPES: ReadonlyMap<string, string> = new Map([
  ["TEXT", "sa.Text()"],
  ["BIGSERIAL", "sa.BigInteger()"],
  ["BIGINT", "sa.BigInteger()"],
  ["INTEGER", "sa.Integer()"],
  ["SERIAL", "sa.Integer()"],
  ["BOOLEAN", "sa.Boolean()"],
  ["DOUBLE PRECISION", "sa.Float()"],
  ["DATE", "sa.Date()"],
  ["TIMESTAMPTZ", "sa.DateTime(timezone=True)"],
  ["UUID", "sa.Uuid()"],
  ["BYTEA", "sa.LargeBinary()"],
  ["JSONB", "postgresql.JSONB()"],
]);

function mapServerDefault(value: string | null): string | null {
  if (value === null) return null;
  if (value === "NOW()") return "sa.func.now()";
  return `sa.text(${pythonStringLiteral(value)})`;
}

function pythonStringLiteral(s: string): string {
  if (!s.includes("'")) return `'${s}'`;
  if (!s.includes('"')) return `"${s}"`;
  return `"${s.replace(/"/g, '\\"')}"`;
}

function renderColumnCall(c: AlembicColumn): string {
  const parts: string[] = [`"${c.name}"`, c.saType];
  if (c.primaryKey) parts.push("primary_key=True");
  if (c.autoincrement) parts.push("autoincrement=True");
  if (c.serverDefault !== null) parts.push(`server_default=${c.serverDefault}`);
  parts.push(`nullable=${c.nullable ? "True" : "False"}`);
  return `sa.Column(${parts.join(", ")})`;
}

function renderForeignKeyCall(fk: AlembicForeignKey): string {
  return `sa.ForeignKeyConstraint(["${fk.column}"], ["${fk.refTable}.${fk.refColumn}"], ondelete="${fk.onDelete}", name="${fk.name}")`;
}

function renderCheckCall(c: AlembicCheck): string {
  return `sa.CheckConstraint(${pythonStringLiteral(c.sql)}, name="${c.name}")`;
}

File: modules/codegen/src/main/scala/specrest/codegen/alembic/Migration.scala
package specrest.codegen.alembic

import specrest.codegen.migration.AlembicSyntax
import specrest.codegen.migration.CanonicalType
import specrest.codegen.migration.Dialect
import specrest.codegen.migration.Postgres
import specrest.codegen.migration.SchemaDiff
import specrest.ir.generated.SpecRestGenerated.*

final case class AlembicColumn(
    name: String,
    saType: String,
    nullable: Boolean,
    primaryKey: Boolean,
    autoincrement: Boolean,
    serverDefault: Option[String]
)

final case class AlembicForeignKey(
    name: String,
    column: String,
    refTable: String,
    refColumn: String,
    onDelete: String
)

final case class AlembicCheck(name: String, sql: String)

final case class AlembicIndex(
    name: String,
    table: String,
    columns: List[String],
    unique: Boolean,
    postgresqlWhereSuffix: String
)

final case class AlembicTrigger(
    name: String,
    upgradeStatements: List[String],
    downgradeStatements: List[String]
)

final case class AlembicTable(
    name: String,
    entityName: String,
    columns: List[AlembicColumn],
    foreignKeys: List[AlembicForeignKey],
    checks: List[AlembicCheck],
    indexes: List[AlembicIndex],
    tableArgs: List[String]
)

final case class AlembicMigration(
    revision: String,
    createdDate: String,
    tables: List[AlembicTable],
    tablesReversed: List[AlembicTable],
    triggers: List[AlembicTrigger],
    triggersReversed: List[AlembicTrigger],
    needsPostgresDialect: Boolean
)

final case class BuildMigrationOptions(
    revision: Option[String] = None,
    createdDate: Option[String] = None
)

object Migration:

  def buildAlembicMigration(
      schema: database_schema,
      opts: BuildMigrationOptions = BuildMigrationOptions(),
      dialect: Dialect = Postgres
  ): AlembicMigration =
    val sorted               = SchemaDiff.topoSort(schemaTables(schema))
    val tables               = sorted.map(buildAlembicTable(_, dialect))
    val triggers             = schemaTriggers(schema).map(buildAlembicTrigger(_, dialect))
    val needsPostgresDialect = tables.exists(_.columns.exists(_.saType.startsWith("postgresql.")))
    AlembicMigration(
      revision = opts.revision.getOrElse("001"),
      createdDate = opts.createdDate.getOrElse(java.time.LocalDate.now.toString),
      tables = tables,
      tablesReversed = tables.reverse,
      triggers = triggers,
      triggersReversed = triggers.reverse,
      needsPostgresDialect = needsPostgresDialect
    )

  private def buildAlembicTrigger(t: trigger_spec, dialect: Dialect): AlembicTrigger =
    val emission = dialect.renderTrigger(t)
    AlembicTrigger(
      name = triggerName(t),
      upgradeStatements = emission.upgrade,
      downgradeStatements = emission.downgrade
    )

  private def buildAlembicTable(t: table_spec, dialect: Dialect): AlembicTable =
    val tname       = tableName(t)
    val columns     = tableColumns(t).map(buildColumn(_, t, dialect))
    val foreignKeys = tableForeignKeys(t).map: fk =>
      AlembicForeignKey(
        name = s"fk_${tname}_${fkColumn(fk)}",
        column = fkColumn(fk),
        refTable = fkRefTable(fk),
        refColumn = fkRefColumn(fk),
        onDelete = fkOnDelete(fk)
      )
    val checks =
      SchemaDiff.namedChecks(t, dialect, SchemaDiff.autoIncrementPk(t)).map: (name, sql) =>
        AlembicCheck(name = name, sql = sql)
    val indexes = tableIndexes(t).map: ix =>
      // Mirror sqlCreateIndex: when a partial index degrades on a dialect without partial-index
      // support, drop UNIQUE too (a full unique index over-enforces). Keeps the Alembic and raw
      // SQL renderers consistent — no cross-renderer schema drift.
      val partialDropped = indexFilterClause(ix).isDefined && !dialect.caps.supportsPartialIndex
      AlembicIndex(
        name = indexName(ix),
        table = tname,
        columns = indexColumns(ix),
        unique = indexUnique(ix) && !partialDropped,
        postgresqlWhereSuffix = dialect.partialIndex(ix).value
      )
    val tableArgs =
      columns.map(renderColumnCall) ++
        foreignKeys.map(renderForeignKeyCall) ++
        checks.map(renderCheckCall)
    AlembicTable(
      name = tname,
      entityName = tableEntityName(t),
      columns = columns,
      foreignKeys = foreignKeys,
      checks = checks,
      indexes = indexes,
      tableArgs = tableArgs
    )

  private def buildColumn(c: column_spec, t: table_spec, dialect: Dialect): AlembicColumn =
    val sqlT     = columnSqlType(c)
    val isPk     = columnName(c) == tablePrimaryKey(t)
    val isSerial = CanonicalType.isAutoIncrementType(sqlT)
    AlembicColumn(
      name = columnName(c),
      saType = AlembicSyntax.mapSqlTypeToSa(sqlT, dialect),
      nullable = columnNullable(c),
      primaryKey = isPk,
      autoincrement = isSerial,
      serverDefault =
        AlembicSyntax.mapServerDefault(columnDefaultValue(c).map(dialect.alembicServerDefault))
    )

  private def renderColumnCall(c: AlembicColumn): String =
    val parts = List.newBuilder[String]
    parts += s""""${c.name}""""
    parts += c.saType
    if c.primaryKey then parts += "primary_key=True"
    // An explicit integer PK is application-supplied (spec `next_id`), not DB-generated. SQLAlchemy
    // defaults a lone integer PK to `autoincrement="auto"` (-> SERIAL/AUTO_INCREMENT), so the
    // non-serial PK must pin `autoincrement=False` to stay consistent with the raw-SQL/Prisma
    // renderers and the spec. Synthesized serial PKs keep `autoincrement=True`.
    if c.primaryKey then parts += s"autoincrement=${if c.autoincrement then "True" else "False"}"
    else if c.autoincrement then parts += "autoincrement=True"
    c.serverDefault.foreach(d => parts += s"server_default=$d")
    parts += s"nullable=${if c.nullable then "True" else "False"}"
    s"sa.Column(${parts.result().mkString(", ")})"

  private def renderForeignKeyCall(fk: AlembicForeignKey): String =
    s"""sa.ForeignKeyConstraint(["${fk.column}"], ["${fk.refTable}.${fk.refColumn}"], ondelete="${fk
        .onDelete}", name="${fk.name}")"""

  private def renderCheckCall(c: AlembicCheck): String =
    s"""sa.CheckConstraint(${AlembicSyntax.pythonStringLiteral(c.sql)}, name="${c.name}")"""
